sinatra-contrib/lib/sinatra/capture.rb

```diff
@@ -84,40 +84,27 @@ module Sinatra
   module Capture
     include Sinatra::EngineTracking
 
-    DUMMIES = {
-      :haml   => "!= capture_haml(*args, &block)",
-      :erubi  => "<% @capture = yield(*args) %>",
-      :erubis => "<% @capture = yield(*args) %>",
-      :slim   => "== yield(*args)"
-    }
-
     def capture(*args, &block)
-      @capture = nil
-      if current_engine == :ruby
-        result = block[*args]
-      elsif current_engine == :erb || current_engine == :slim
-        @_out_buf, _buf_was = '', @_out_buf
-        raw = block.call(*args)
-        captured = block.binding.eval('@_out_buf')
-        result = captured.empty? ? raw : captured
-        @_out_buf = _buf_was
+      if ruby?
+        block[*args]
+      elsif haml?
+        buffer = Haml::Buffer.new(nil, Haml::Options.new.for_buffer)
+        with_haml_buffer(buffer) { capture_haml(*args, &block) }
       else
-        buffer     = eval '_buf if defined?(_buf)', block.binding
-        old_buffer = buffer.dup if buffer
-        dummy      = DUMMIES.fetch(current_engine)
-        options    = { :layout => false, :locals => {:args => args, :block => block }}
-
-        buffer.clear unless buffer.nil?
-        result = render(current_engine, dummy, options, &block)
```

Comment:
Should we use the `render` method for reproducing template environment? If not, could you explain why this isn't needed?

Reply:
I've added comment https://github.com/sinatra/sinatra/pull/1502#issuecomment-447656203 to  explain why I don't think using render is necessary.